engine/main.ml

```diff
@@ -224,39 +230,337 @@ let rec parse_lambda lsp =
+      )
+  in
+  let bprintf = Printf.bprintf
+  in
+  let rec bprint_accessor buf = function
+    | AcRoot (v) -> bprintf buf "AcRoot=%s" v
+    | AcField (a, i) -> bprintf buf "AcField(%d %a)" i bprint_accessor a
+    | AcTag (a, i) -> bprintf buf "AcTag(%d %a)" i bprint_accessor a
+  in
+  let bprint_pi buf pi =
+    let print_op buf = function
+      | Tag i -> bprintf buf "Tag %d" i
+      | NotTag i -> bprintf buf "NotTag %d" i
+      | Int i-> bprintf buf "Int %d" i
+      | NotInt i-> bprintf buf "NotInt %d" i
+      | Ge i -> bprintf buf "Ge %d" i
+      | Gt i -> bprintf buf "Gt %d" i
+      | Le i -> bprintf buf "Le %d" i
+      | Lt i -> bprintf buf "Lt %d" i
+      | Eq i -> bprintf buf "Eq %d" i
+      | Nq i -> bprintf buf "Nq %d" i
+      | Isout i -> bprintf buf "Isout %d" i
+      | Isin i -> bprintf buf "Isin %d" i
+    in
+    bprintf buf "{ var=%a; op=%a; }" bprint_accessor pi.var print_op pi.op
+  in
+  let rec bprint_list ~sep bprint buf = function
+    | [] -> ()
+    | [x] -> bprint buf x
+    | x :: xs ->
+       bprintf buf "%a%t%a"
+         bprint x
+         sep
+         (bprint_list ~sep bprint) xs in
+  let break ntabs buf =
+    bprintf buf "\n%s" (BatList.init ntabs (fun _ -> "\t") |> String.concat "") in
+  let rec bprint_tree ntabs buf tree =
+    let sep = break ntabs in
+    match tree with
+    | Leaf (pilist, target_blackbox) ->
+       bprintf buf
+         "Leaf=%s\
+              %t%a"
+         target_blackbox
+         (break ntabs) (bprint_list ~sep bprint_pi) pilist
+    | Node (cst_list) ->
+       bprintf buf
+         "Node=\
+          %t%a"
+         (break ntabs)
+         (bprint_list ~sep (bprint_tree (ntabs+1))) cst_list
+  and
+  bprint_env ntabs buf env =
+    let bprint_binding buf (key, entry) =
+      let bprint_value buf = function
+        | Accessor a -> bprint_accessor buf a
+        | Addition (v, i) ->
+           bprintf buf "Addition=%s,%i" v i
+        | Function (v, f_tree) ->
+           bprintf buf  "Function=%s,ConstraintTree: %a"
+             v
+             (bprint_tree (ntabs + 1)) f_tree
+        | Catch (e, vars, c_tree) ->
+          bprintf buf "Catch=%d %s,ConstraintTree: %a"
+            e
+            (String.concat " " vars)
+             (bprint_tree (ntabs + 1)) c_tree
+      in
+      bprintf buf "%s: %a" key bprint_value entry
+    in
+    bprint_list ~sep:(break ntabs) bprint_binding buf (SMap.bindings env)
+  in
+  let print_env env =
+    let buf = Buffer.create 42 in
+    bprint_env 0 buf env;
+    BatIO.write_line BatIO.stdout (Buffer.contents buf)
+  in
+  let expand_env variable entry : environment SMap.t = 
+    let _ = match SMap.find_opt variable env with
+      | Some _ -> assert false
+      | None -> ()
+    in
+    SMap.add variable entry env
+  in
+  (* perform union on two maps, keys should never differ *)
+  let union env1 env2 = SMap.union (fun _ a b -> assert (a = b); Some a) env1 env2
+  in
+  let match_let_accessor env acc =
```

Comment:
This function name is not so clear. From the code it looks like this function is used to evaluate `let` bindings. Maybe `eval_let_definition`? Is it just the generic code to evaluation an expression? (`eval_expr`?)

Reply:
We discussed this and @FraMecca suggests making the grammar/parser/types stricter to clarify that let definitions are not arbitrary `sexpr` but a smaller subset. It's a good change to make, probably after this PR.